network/src/peer/mod.rs

```diff
@@ -0,0 +1,426 @@
+                            warn!("Failed to send notification to DirectSend actor. Error: {:?}", err);
+                        }
+                      } else {
+                        unreachable!("Negotiated unsupported protocol");
+                    }
+                  }
+                  Err(e) => {
+                    error!(
+                      "Inbound substream negotiation for peer {} failed: {}",
+                      self.identity.peer_id().short_str(), e
+                    );
+                  }
+                }
+              },
+              _ = pending_outbound_substreams.select_next_some() => {
+                // Do nothing since these futures have an output of "()"
+              },
+            }
+        }
+        debug!(
+            "Peer actor '{}' shutdown",
+            self.identity.peer_id().short_str()
+        );
+    }
+
+    async fn handle_request<'a>(
+        &'a mut self,
+        pending: &'a mut FuturesUnordered<BoxFuture<'static, ()>>,
+        request: PeerRequest<TMuxer::Substream>,
+    ) {
+        trace!(
+            "Peer {} PeerRequest::{:?}",
+            self.identity.peer_id().short_str(),
+            request
+        );
+        match request {
+            PeerRequest::OpenSubstream(protocol, channel) => {
+                pending.push(self.handle_open_outbound_substream_request(protocol, channel));
+            }
+            PeerRequest::CloseConnection => {
+                self.close_connection(DisconnectReason::Requested).await;
+            }
+        }
+    }
+
+    fn handle_open_outbound_substream_request(
+        &self,
+        protocol: ProtocolId,
+        channel: oneshot::Sender<Result<TMuxer::Substream, PeerManagerError>>,
+    ) -> BoxFuture<'static, ()> {
+        let outbound = self.connection.open_outbound();
+        let optimistic_negotiation = self.identity.is_protocol_supported(&protocol);
+        let negotiate = Self::negotiate_outbound_substream(
+            self.identity.peer_id(),
+            outbound,
+            protocol,
+            optimistic_negotiation,
+            channel,
+        );
+
+        negotiate.boxed()
+    }
+
+    async fn negotiate_outbound_substream(
+        peer_id: PeerId,
+        outbound_fut: TMuxer::Outbound,
+        protocol: ProtocolId,
+        optimistic_negotiation: bool,
+        channel: oneshot::Sender<Result<TMuxer::Substream, PeerManagerError>>,
+    ) {
+        let response = match outbound_fut.await {
+            Ok(substream) => {
+                // TODO(bmwill) Evaluate if we should still try to open and negotiate an outbound
+                // substream even though we know for a fact that the Identity struct of this Peer
+                // doesn't include the protocol we're interested in.
+                if optimistic_negotiation {
+                    negotiate_outbound_select(substream, &protocol).await
+                } else {
+                    warn!(
```

Comment:
right now we're doing interactive negotiation for the initial connection handshake and then optimistic negotiation for all protocols after we finish the Identity/supported protocols list exchange.

you use interactive negotiation when you don't know what protocols the other side supports. for example, all I know when I dial a peer today is "/ip4/1.2.3.4/tcp/567", their peer id, and their noise identity pubkey. I don't know what version of noise they support, what version of yamux, what version of identity protocol, etc.... So we have to interactively negotiate like:

```
> (tcp handshake)
< (tcp handshake)

> [tcp] "hey, let's talk /noise_ix/1.0.0"
< [tcp] "sorry, not supported"

> [tcp] "hey, let's talk /noise_ix/2.0.0"
< [tcp] "ok, I support /noise_ix/2.0.0"
> [tcp] (noise handshake)
< [tcp] (noise handshake)

> [tcp+noise] "hey let's talk /yamux/1.0.0"
< [tcp+noise] "ok, I support /yamux/1.0.0"

> [tcp+noise+yamux] "hey let's talk /identity/1.0.0"
< [tcp+noise+yamux] "ok I support /identity/1.0.0"
> [tcp+noise+yamux] identity: here's my list of supported protocols: /rpc/consensus/1, /direct-send/mempool/2, etc...
< [tcp+noise+yamux] identity: here's my list of supported protocols: /rpc/consensus/1, /direct-send/mempool/4, etc...
```

now that we've exchanged our full list of protocols, we can just prefix any message with the protocol we're going to speak and know that a normal peer will accept it if he advertises it as supported. we're calling this "optimistic negotiation":

```
> this message protocol is: "/direct-send/consensus/1", the message is: "Proposal { ... }"
```

Of course, we should absolutely not negotiate Noise protocol in-the-clear, since it leaves us open to various protocol downgrading attacks. This issue is tracked here: https://github.com/libra/libra/issues/1158

Reply:
Thanks for the great explanation. Is the logging level appropriate for something that happens when we establish a connection?